### ISSUE TITLE
Incorrect Line Number Assigned to "Unexpected Token" Parse Errors

### ISSUE THREAD

**Issue by bdkjones** [eslint/eslint#10077]
* **ESLint Version:** 4.18.2
* **Node Version:** 8.9.4
* **npm Version:** 5.6.0

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**
N/A - no ESLint configuration is needed for this issue; see below.

</details>

**Summary**
When ESLint encounters a parse error with an "unexpected token", it incorrectly assigns the line number of the error to the final line of the file instead of the line where the unexpected token was actually encountered.

**Source Code Producing The Issue**
NOTE: the following source code should contain **12** lines when copied to an editor. Lines 10 and 11 are newline-terminated, leaving line 12 as the next available line. See the screenshot at bottom.
```js
( function() {
	$( 'body' ).click( function( e ) 
	{
		var clickedOn = $( e.target );
		if ( clickedOn.parents().andSelf().is( '#foo' ) ) {
			// Do something.
		} else {
			// Do nothing.
		}
	} );

```

```bash
eslint --no-eslintrc test.js
```

**What did you expect to happen?**

The source code above is missing a brace on line 10. ESLint should report "parse error, unexpected token on line 10". 

**What actually happened?**

ESLint reports the parse error and unexpected token as occurring on line 12 instead of line 10. The raw output from ESLint is:

    12:1  error  Parsing error: Unexpected token

Here the difference is only two lines, but if we add more lines to the end of the file, the difference grows because ESLint always assigns the unexpected token error to the last line of the file.

Or, shown here:

<img width="1003" alt="screen shot 2018-03-09 at 18 23 36" src="https://user-images.githubusercontent.com/926566/37237536-2db1ba30-23c9-11e8-8049-174b8d049b8c.png">

**Comment by bdkjones** [comment]
This appears to manifest even when there is content beyond the line where the parse error occurs, as shown here:

<img width="985" alt="screen shot 2018-03-09 at 18 45 28" src="https://user-images.githubusercontent.com/926566/37237608-5e9bd422-23ca-11e8-9073-2205a2d5661d.png">

**Comment by not-an-aardvark** [comment]
Thanks for the report. However, I'm confused -- couldn't the missing brace go anywhere after line 10, including at the beginning of line 12? I assume what happened was that the parser scanned until the end of the file looking for a brace, and reported an error when it couldn't find one.

**Comment by bdkjones** [comment]
That is true. But I think a better approach is to report the line number where the missing token actually is, rather than the line number where the parser gave up looking for it. 

The former line number is much more useful context for a human because that's the line we should look at to see what's missing. In this simplified case, it's easy to see the problem. In more complex cases, however, the reported line number could be a ways off. 

Accurately reporting the line number is also very helpful for tools that integrate ESLint. For example, in the case above, an editor plugin would want to highlight line 11 and show a "missing brace". With more empty space, ESLint reports an ever-higher line number and the highlighting is therefore off, drawing the eye away from where the real issue lies.

In short, the current approach is excellent for the parser but bad for humans.

**Comment by not-an-aardvark** [comment]
> But I think a better approach is to report the line number where the missing token actually is, rather than the line number where the parser gave up looking for it.

I'm not sure I agree -- the missing token isn't anywhere! It's missing. 😃 

In any case, I would recommend reporting this issue on the [`acorn`](https://github.com/acornjs/acorn) repository instead, since that's the library that's actually doing the parsing here. ESLint is just passing on the error location that was originally provided by `acorn`.

**Comment by bdkjones** [comment]
Okay, thanks!

I guess another way to consider it is: suppose this file had 9 blank lines at the end. Where would a human insert the missing brace to fix the problem? Line 19? Unlikely. They'll choose line 11. 

ESLint is going to complain about a missing token on line 19. But when you see that, your first reaction is, "Why is it complaining about line 19? There's nothing even close to that."

By contrast, if ESLint reports the *first* spot that the missing token could be rather than the last spot, you get a line number that's much closer to the actual problem AND makes it faster for a human to grok what's going on—the original function block isn't closed properly.

**Comment by platinumazure** [comment]
> By contrast, if ESLint reports the first spot that the missing token could be rather than the last spot

You mean right after the opening token, on the same line, right?

No? I guess that means you want it reported *somewhere* in between the first possible location and the last possible location, but I pity any developer who tries to solve this general problem. :smile:

I could see an argument for reporting at the opening token (Missing matching `}` for `{` token at line 10), though.

**Comment by bdkjones** [comment]
Changing the message from a generic “parse error: unknown token” to “missing closing } for opening { on line 1” would be utterly fantastic if that’s achievable. Talk about an improvement.

**Comment by bdkjones** [comment]
Here's a better case showing that the current behavior is very misleading. Let's run ESLint on this file:

```
( function() {
	$( 'body' ).click( function( e ) 
	{
		var clickedOn = $( e.target );
		if ( clickedOn.parents().andSelf().is( '#foo' ) ) {
			// Do something.
		} else {
			// Do nothing.
		}
	} );
// This line should contain '});', but it's missing.

( function() {
	$( 'body' ).click( function( e ) 
	{
		var clickedOn = $( e.target );
		if ( clickedOn.parents().andSelf().is( '#foo' ) ) {
			// Do something.
		} else {
			// Do nothing.
		}
	});
});

( function() {
	$( 'body' ).click( function( e ) 
	{
		var clickedOn = $( e.target );
		if ( clickedOn.parents().andSelf().is( '#foo' ) ) {
			// Do something.
		} else {
			// Do nothing.
		}
	} );
});

```

The parse error is on line 11. That's where a human would insert `});` to fix the problem. But if you run ESLint on this, it's going to report an unexpected token on **line 36**. That's not even close to where the actual problem is.

**Comment by not-an-aardvark** [comment]
Again, I would recommend reporting this issue on the [`acorn`](https://github.com/acornjs/acorn) repository instead, since that's the library that's actually doing the parsing here. ESLint is just passing on the error location that was originally provided by `acorn`.

**Comment by bdkjones** [comment]
Done. Just wanted to update this issue with a better example.

**Comment by not-an-aardvark** [comment]
Closing this issue since there's not anything we can change in `eslint`. If this gets changed in `acorn`, then `eslint`'s users will be able to take advantage of it automatically.
